src/register_target.cpp

```diff
@@ -33,42 +34,88 @@ inline namespace MIGRAPHX_INLINE_NS {
 void store_target_lib(const dynamic_loader& lib)
 {
     static std::vector<dynamic_loader> target_loader;
+    static std::mutex mutex;
+    std::unique_lock<std::mutex> lock(mutex);
+
     target_loader.emplace_back(lib);
 }
 
+/**
+ * Returns a singleton map of targets and names.
+ */
 std::unordered_map<std::string, target>& target_map()
 {
     static std::unordered_map<std::string, target> m; // NOLINT
     return m;
 }
 
+/**
+ * Returns a singleton mutex used by the various register_target methods.
+ */
+std::mutex& target_mutex()
+{
+    static std::mutex m; // NOLINT
+    return m;
+}
+
 void register_target_init() { (void)target_map(); }
 
 void unregister_target(const std::string& name)
 {
+    std::unique_lock<std::mutex> lock(target_mutex());
     assert(target_map().count(name));
     target_map().erase(name);
 }
 
-void register_target(const target& t) { target_map()[t.name()] = t; }
+/**
+ * Insert a target name in the target_map; thread safe.
+ */
+void register_target(const target& t)
+{
+    std::unique_lock<std::mutex> lock(target_mutex());
+    target_map()[t.name()] = t;
```

Comment:
You should be unlocking here.

Reply:
With std::unique_lock, the destructor automatically unlocks it when it goes out of scope.  This is supposed not only to be a coding convenience, but gives protection when things are interrupted by an exception. From [cplusplus.com](https://cplusplus.com/reference/mutex/unique_lock/):
"This class guarantees an unlocked status on destruction (even if not called explicitly). Therefore it is especially useful as an object with automatic duration, as it guarantees the mutex object is properly unlocked in case an exception is thrown."